Audit branch

Evan-Kim2028 opened the pull request:
* Refactored the env variables and flags to use [urfav cli v2 library](https://cli.urfave.org/).
* Refactored repository to reorganize around main.go and everythig else in "internal" package
* removed self creation of hTTP client
* check return body contents in `sendBundle` function
* replaced `context.Background()` with `context.WithTimeout`
* added contract addresses as env variables
*

Evan-Kim2028 committed:
replace env variables and flags with urfav cli

Evan-Kim2028 committed:
add main.go to top level and remove cmd directory

Evan-Kim2028 committed:
rename core to internal, reorganize folder directory

Evan-Kim2028 committed:
update golang

Evan-Kim2028 committed:
replace context background with context timeout

Evan-Kim2028 committed:
remove http client and use constant timeout variable to set timeout

Evan-Kim2028 committed:
check body contents when function gets returned

Evan-Kim2028 committed:
load environment variables before the cli app

Evan-Kim2028 committed:
remove unused logging module

Evan-Kim2028 committed:
add priority fee

Evan-Kim2028 committed:
feat: turn the addresses into environmental variables

Evan-Kim2028 committed:
remove hard coded chain ID

Evan-Kim2028 committed:
remove timer and have for loop run indefinitely

Evan-Kim2028 committed:
update example .env file

Evan-Kim2028 committed:
update example .env file

Evan-Kim2028 committed:
fix: linting

Evan-Kim2028 committed:
refactor: move client functions from main to client

Evan-Kim2028 committed:
refactor: move sendpreconfbid from main to bidderapi

Evan-Kim2028 committed:
refactor: replace eth_transfer and blob .env variables with single num_blob variable

Evan-Kim2028 committed:
feat: add priority fee to blob

Evan-Kim2028 committed:
fix: change num blob from Uint to Uint64

Evan-Kim2028 committed:
refactor: switch logging to zerolog

Evan-Kim2028 committed:
docs: update .env example

Evan-Kim2028 committed:
docs: update readme

Evan-Kim2028 committed:
refactor: make defaultTimeout a .env variable

Evan-Kim2028 committed:
docs: update comments

Evan-Kim2028 committed:
chore: linting

Evan-Kim2028 committed:
feat: add valid pubkey check

Evan-Kim2028 committed:
feat: add unit testing for SendPreconfBid

Evan-Kim2028 committed:
fix: add nil transaction check

Evan-Kim2028 committed:
feat: add sendBid test

Evan-Kim2028 committed:
refactor: modularize sendbid function

Evan-Kim2028 committed:
feat: add SendBid success and request error tests

Evan-Kim2028 committed:
docs: update with test command

Evan-Kim2028 commented:
* Added unit tests coverage for `bidderapi.go` functions `sendPreconfBid` and `sendBid`. These are the most important functions and will likely change the most over time in future releases
* switched geth logging for zerolog logging
* refactored environment variables for more simplicity, simplifed main.go logic
* refactored sendBid function to be more modular

Evan-Kim2028 committed:
feat: add logging for preconf bid params

mrekucci commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM golang:1.21
+FROM golang:1.22
```
Use Golang 1.23

mrekucci commented on go.mod:
```diff
@@ -1,12 +1,13 @@
 module github.com/primev/preconf_blob_bidder
 
-go 1.21
+go 1.22
```
Use go 1.23

mrekucci commented on go.mod:
```diff
@@ -1,12 +1,13 @@
-go 1.21
+go 1.22
+
+toolchain go1.23.0
```
This shouldn't be needed.

mrekucci commented on go.mod:
```diff
@@ -15,36 +16,46 @@ require (
 require (
```
This require block can be merged with the one above.

mrekucci commented on go.mod:
```diff
@@ -15,36 +16,46 @@ require (
+	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
 
 require (
```
This require block can be merged with the one above.

mrekucci commented on internal/eth/bundle.go:
```diff
@@ -0,0 +1,132 @@
+	Error  *RPCError       `json:"error"`
+}
+
+type RPCError struct {
```
This can be embedded into `JSONRPCResponse`.

mrekucci commented on internal/eth/bundle.go:
```diff
@@ -0,0 +1,132 @@
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/rs/zerolog/log"
```
Use `log/slog` from standard library across the whole codebase.

mrekucci commented on internal/eth/bundle.go:
```diff
@@ -0,0 +1,132 @@
+		return "", err
+	}
+
+	// Create a context with a timeout
```
This comment (and the similar one above and below) adds no value. If necessary, the comment should include an explanation of why something is happening, not what is happening (the what should be obvious from the code).

mrekucci commented on internal/eth/bundle.go:
```diff
@@ -0,0 +1,132 @@
+	ID      int                      `json:"id"`
+}
+
+func SendBundle(RPCURL string, signedTx *types.Transaction, blkNum uint64) (string, error) {
```
`RPCURL` -> `rpcurl`

mrekucci commented on internal/eth/bundle.go:
```diff
@@ -0,0 +1,132 @@
+	defer cancel()
+
+	// Create a new HTTP POST request with the JSON payload
+	req, err := http.NewRequestWithContext(ctx, "POST", RPCURL, bytes.NewBuffer(payloadBytes))
```
`"POST"` -> `http.MethodPost`
`bytes.NewBuffer(payloadBytes)` -> `bytes.NewReader(payloadBytes)`

mrekucci commented on internal/eth/bundle.go:
```diff
@@ -0,0 +1,132 @@
+			Int("code", rpcResp.Error.Code).
+			Str("message", rpcResp.Error.Message).
+			Msg("Received error from RPC")
+		return "", fmt.Errorf("RPC Error %d: %s", rpcResp.Error.Code, rpcResp.Error.Message)
```
`"RPC Error"`-> `"request failed:"

mrekucci commented on internal/eth/sendtx.go:
```diff
@@ -18,88 +21,164 @@ import (
+
+// init initializes the defaultTimeout variable by reading the DEFAULT_TIMEOUT
+// environment variable. If not set or invalid, it defaults to 15 seconds.
+func init() {
```
This is unnecessary, it should be set to it's default value where the flag is defined.

Evan-Kim2028 replied:
The default timeout flag is defined in the main.go() file as a .env variable. The defaultTimeout variable is also used in the bundle.go file as well as sendtx.go.

This was the best I could come up with to use the flag defined value without refactoring the functions to require a defaulttimeout parameter

mrekucci replied:
I would suggest creating a service struct in which you set all the necessary options that are known at the time of creation and will not change (e.g. default timeout, rpcurl, etc.) and fields that must maintain some state (e.g. logger, ethclient, httpclient, etc.). Then attach existing functions to this structure. This will result in fewer parameters in these methods and avoid global state. Your code will also be easier to mock and test.

mrekucci commented on internal/eth/sendtx.go:
```diff
@@ -18,88 +21,164 @@ import (
+	blockNumber := header.Number.Uint64()
+
+	// Create a transaction with a priority fee.
+	priorityFee := big.NewInt(2_000_000_000) // 2 gwei in wei
```
This two values can be passed directly to `types.DynamicFeeTx` without defining them here.

Evan-Kim2028 replied:
`blockNumber` gets printed out later as a log. Is it better to pass `header.Number.Uint64()` value directly to both `types.DynamicFeeTx` and the log?

Similarly `priorityFee` value is used to calculate `maxFee`. Is it better to pass the maxFee directly into `types.DynamicFeeTx` as well?

mrekucci commented on internal/eth/sendtx.go:
```diff
@@ -119,36 +198,52 @@ func ExecuteBlobTransaction(client *ethclient.Client, authAcct bb.AuthAcct, numB
 	maxFeePerGas := baseFee
-
+	// Use for nonzero priority fee
+	priorityFee := big.NewInt(5_000_000_000) // 5 gwei in wei
```
The same as above.

mrekucci commented on internal/mevcommit/client.go:
```diff
@@ -0,0 +1,289 @@
+// - A pointer to an ethclient.Client for interacting with the Ethereum node, or an error if the connection fails.
+func NewGethClient(endpoint string) (*ethclient.Client, error) {
+	// Dial the Ethereum RPC endpoint
+	client, err := rpc.Dial(endpoint)
```
Isn't there a `Dial` method with context?

mrekucci commented on internal/mevcommit/client.go:
```diff
@@ -0,0 +1,289 @@
+				Str("ws_endpoint", MaskEndpoint(wsEndpoint)).
+				Int("attempt", i+1).
+				Msg("WebSocket client reconnected")
+			sub, err = wsClient.SubscribeNewHead(context.Background(), headers)
```
There should be context with timeout instead of `context.Background()`.

mrekucci commented on internal/mevcommit/client.go:
```diff
@@ -0,0 +1,289 @@
+//
+// Returns:
+// - A masked version of the endpoint if its length exceeds 10 characters, otherwise a fixed mask.
+func MaskEndpoint(endpoint string) string {
```
This is unnecessary, if you use `log/slog`, you can redact a secret as follows: https://pkg.go.dev/golang.org/x/exp/slog#example-LogValuer-Secret

Evan-Kim2028 replied:
I think I will just leave this as is and add an enchancement issue to track for now

mrekucci commented on main.go:
```diff
@@ -0,0 +1,300 @@
+
+func main() {
+	// Load the .env file before setting up the app
+	envFile := os.Getenv("ENV_FILE")
```
This is not needed, since the `ENV_FILE` is declared in the cli flags below.

mrekucci commented on main.go:
```diff
@@ -0,0 +1,300 @@
+		Name:  "Preconf Bidder",
+		Usage: "A tool for bidding in mev-commit preconfirmation auctions for blobs and transactions",
+		Flags: []cli.Flag{
+			&cli.StringFlag{
```
If you export all flags to the `var` block, you can refer to them by their name instead of the string in the name.

mrekucci commented on main.go:
```diff
@@ -0,0 +1,300 @@
+
+			cfg := bb.BidderConfig{
+				ServerAddress: bidderAddress,
+				LogFmt:        "json",
```
What are these `Log*` files for when thy are set but not used?

Evan-Kim2028 replied:
removed

mrekucci commented on main.go:
```diff
@@ -0,0 +1,300 @@
+					var blockNumber uint64
+					if numBlob == 0 {
+						// Perform ETH Transfer
+						amount := new(big.Int).SetInt64(1e15)
```
`new(big.Int).SetInt64(1e15)` -> `big.NewInt(1e15)`

mrekucci commented on main.go:
```diff
@@ -0,0 +1,300 @@
+					randomEthAmount := rand.NormFloat64()*stdDev + bidAmount
+
+					// Ensure the randomEthAmount is positive
+					if randomEthAmount <= 0 {
```
You can use `max(randomEthAmount, bidAmount)`.

mrekucci commented on main.go:
```diff
@@ -0,0 +1,300 @@
+}
+
+// loadEnvFile loads the specified .env file into the environment variables
+func loadEnvFile(filename string) error {
```
Consider using: https://github.com/joho/godotenv

Evan-Kim2028 committed:
chore; bump go version, reorganize go mod requirements

Evan-Kim2028 committed:
refactor: SendBundle error responses

Evan-Kim2028 committed:
fix: add context timeouts

Evan-Kim2028 committed:
refactor: replace zerolog with standard library logging

Evan-Kim2028 committed:
fix: remove uneceessary log config

Evan-Kim2028 committed:
refactor: integer type and positive integer logic

Evan-Kim2028 commented:
* used more context timeout modifications
* changed logging from zerolog to standard library logging
* refactor environment variable handling
* updated go mod and version
* refactor minor formatting and reduce logic complexity

mrekucci commented on Dockerfile:
```diff
@@ -7,10 +7,7 @@ RUN go mod download
-ENTRYPOINT ["./getPreconf"]
-
-CMD ["--ethtransfer"]
+ENTRYPOINT ["./preconf_bot"]
```
A new line is missing at the end of this file.

mrekucci commented on internal/eth/bundle.go:
```diff
@@ -0,0 +1,140 @@
+	"fmt"
+	"io"
+	"net/http"
+
```
Unnecessary new line. Also, Imports are not sorted.

mrekucci commented on internal/eth/bundle.go:
```diff
@@ -0,0 +1,140 @@
+
+type JSONRPCResponse struct {
+	Result    json.RawMessage `json:"result"`
+	RPCError  RPCError         `json:"error"`
```
The fields are not formatted properly by gofmt.

mrekucci commented on main.go:
```diff
@@ -0,0 +1,272 @@
+)
+
+// Define flag names as constants
+const (
```
Instead of this, I'd suggest doing the following:
```Golang
var (
    bidderAddressFlag := &cli.StringFlag{
        Name:    "bidder-address",
        Usage:   "Address of the bidder",
        EnvVars: []string{"BIDDER_ADDRESS"},
        Value:   "mev-commit-bidder:13524",
     },
     // other flags...
)

// Now below you can do:
c.String(bidderAddressFlag.Name)

mrekucci commented on main.go:
```diff
@@ -0,0 +1,272 @@
+			&cli.StringFlag{
+				Name:    FlagEnv,
+				Usage:   "Path to .env file",
+				EnvVars: []string{"ENV_FILE"},
```
Consider using some sort of prefix (maybe `PRECONF_BOT`) for all env variables to avoid collisions or so that they already exist.

mrekucci commented on main.go:
```diff
@@ -0,0 +1,272 @@
+				Value:   15, // Default to 15 seconds
+			},
+		},
+		Action: func(c *cli.Context) error {
```
The `c.Context` should be used as parent context in all places where you create a new context. This will help you shutdown the app gracefully.

mrekucci commented on internal/mevcommit/client.go:
```diff
@@ -0,0 +1,300 @@
+//
+// Returns:
+// - A pointer to an ethclient.Client if successful, or nil if all retries fail.
+func ConnectRPCClientWithRetries(rpcEndpoint string, maxRetries int, timeout time.Duration) *ethclient.Client {
```
This function should also return an error as do the functions above.

mrekucci commented on internal/mevcommit/client.go:
```diff
@@ -0,0 +1,300 @@
+//
+// Returns:
+// - A pointer to an ethclient.Client and an ethereum.Subscription if successful, or nil values if all retries fail.
+func ReconnectWSClient(wsEndpoint string, headers chan *types.Header) (*ethclient.Client, ethereum.Subscription) {
```
The same as above.

Evan-Kim2028 committed:
fix: add error returns for functions

Evan-Kim2028 committed:
fix: handle error return value from functions

Evan-Kim2028 committed:
refactor: add service struct for better variable management

Evan-Kim2028 committed:
docs: add code doc comment

Evan-Kim2028 committed:
refactor: add service struct and function methods to struct

Evan-Kim2028 committed:
refactor: slog to use service struct logger

Evan-Kim2028 committed:
refactor: consolidate client management to Service struct and assign functions to struct. Consolidate mevcommit packages into service folder too
